### PR TITLE
[GLUTEN-5476][CH] Triger merge on insert task

### PR DESCRIPTION
**Pull request by loneylee**
## What changes were proposed in this pull request?

(Fixes: \#5476)

## How was this patch tested?

Test by ut

### CODE REVIEW

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/ClickhouseOptimisticTransaction.scala [hunk 1]
```diff
@@ -119,6 +120,16 @@ class ClickhouseOptimisticTransaction(
           }.toMap
       }
 
+      spark.conf.getAll.foreach(
+        entry => {
+          if (
+            entry._1.startsWith(s"${CHBackendSettings.getBackendConfigPrefix}.runtime_settings")
+            || entry._1.equalsIgnoreCase(DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE.key)
```
Comment (zzcclp) [inline backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/ClickhouseOptimisticTransaction.scala:127]: it's better use the 'spark.databricks.delta.optimize.minFileSize' to control the size which will be merged.

Comment (loneylee) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/ClickhouseOptimisticTransaction.scala]: fixed

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/ClickhouseOptimisticTransaction.scala [hunk 2]
```diff
@@ -119,6 +120,16 @@ class ClickhouseOptimisticTransaction(
           }.toMap
       }
 
+      spark.conf.getAll.foreach(
+        entry => {
+          if (
+            entry._1.startsWith(s"${CHBackendSettings.getBackendConfigPrefix}.runtime_settings")
```
Comment (liuneng1994) [inline backends-clickhouse/src/main/scala/org/apache/spark/sql/delta/ClickhouseOptimisticTransaction.scala:126]: redundant blank line

File: cpp-ch/local-engine/Common/CHUtil.cpp [hunk 3]
```diff
@@ -594,10 +566,21 @@ void BackendInitializerUtil::initEnvs(DB::Context::ConfigurationPtr config)
         spark_user = spark_user_c_str;
 }
 
+DB::Field BackendInitializerUtil::toField(String key, String value)
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Common/CHUtil.cpp:569]: use const reference

File: cpp-ch/local-engine/Common/CHUtil.h [hunk 4]
```diff
@@ -32,6 +32,10 @@
 
 namespace local_engine
 {
+static const std::unordered_set<String> bool_value_settings{"mergetree.merge_after_insert"};
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Common/CHUtil.h:35]: 常量用大写字母

File: cpp-ch/local-engine/Storages/CustomStorageMergeTree.cpp [hunk 5]
```diff
@@ -158,14 +160,14 @@ MergeTreeData::MutableDataPartsVector CustomStorageMergeTree::loadDataPartsWithN
         data_parts.emplace_back(res.part);
     }
 
-    if(getStorageID().hasUUID())
-    {
+    // if(getStorageID().hasUUID())
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/CustomStorageMergeTree.cpp:163]: remove useless comments

File: cpp-ch/local-engine/Storages/Mergetree/MetaDataHelper.cpp [hunk 6]
```diff
@@ -142,5 +147,60 @@ void saveFileStatus(
...
+std::vector<MergeTreeDataPartPtr> mergeParts(
+    std::vector<DB::DataPartPtr> selected_parts,
+    const String & new_part_uuid,
+    CustomStorageMergeTreePtr storage,
+    const String  & partition_dir,
+    const String & bucket_dir)
+{
+    // UUID tmp_table_uuid = UUIDHelpers::generateV4();
+    // auto storage = MergeTreeRelParser::parseStorage(
+    //     table, local_engine::SerializedPlanParser::global_context, tmp_table_uuid);
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/Mergetree/MetaDataHelper.cpp:164]: remove

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.h [hunk 7]
```diff
@@ -40,63 +42,56 @@ struct PartInfo
...
-    writeTempPart(DB::BlockWithPartition & block_with_partition, const DB::StorageMetadataPtr & metadata_snapshot);
+    void
+    writeTempPart(MergeTreeDataWriter::TemporaryPart & temp_part, DB::BlockWithPartition & block_with_partition, const DB::StorageMetadataPtr & metadata_snapshot);
     DB::MergeTreeDataWriter::TemporaryPart
     writeTempPartAndFinalize(DB::BlockWithPartition & block_with_partition, const DB::StorageMetadataPtr & metadata_snapshot);
+    void checkAndMerge(bool force = false);
+    void safeEmplaceBackPart(DB::MergeTreeDataPartPtr);
+    void safeAddPart(DB::MergeTreeDataPartPtr);
+    void manualFreeMemory(size_t before_write_memory);
+
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.h:74]: redundant blank line

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.h [hunk 8]
```diff
@@ -40,63 +42,56 @@ struct PartInfo
...
-    DB::MergeTreeData & storage;
+    CustomStorageMergeTreePtr storage;
     DB::StorageMetadataPtr metadata_snapshot;
     DB::ContextPtr context;
     std::unique_ptr<DB::SquashingTransform> squashing_transform;
     int part_num = 1;
-    std::vector<DB::MergeTreeDataPartPtr> new_parts;
+    ConcurrentDeque<DB::MergeTreeDataPartPtr> new_parts;
+
+
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.h:86]: redundant blank line

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp [hunk 9]
```diff
@@ -42,6 +54,55 @@ Block removeColumnSuffix(const DB::Block & block)
...
+    , metadata_snapshot(metadata_snapshot_)
+    , context(context_)
+    , uuid(uuid_)
+    , partition_dir(partition_dir_)
+    , bucket_dir(bucket_dir_)
+    , thread_pool(
+          CurrentMetrics::LocalThread,
+          CurrentMetrics::LocalThreadActive,
+          CurrentMetrics::LocalThreadScheduled,
+          context->getSettings().max_threads,
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:74]: need a new configuration, max_threads is used to limit Select queries

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**